Wrappers/Python/cil/optimisation/utilities/StepSizeMethods.py

```diff
@@ -138,3 +138,88 @@ def get_step_size(self, algorithm):
             raise ValueError(
                 'Could not find a proper step_size in {} loops. Consider increasing alpha or max_iterations.'.format(self.max_iterations))
         return self.alpha
+
+
+class BarzilaiBorweinStepSizeRule(StepSizeRule):
+
+    """ 
+    Applies the Barzilai- Borwein rule to calculate the step size (step_size).
+
+    Let :math:`\Delta x=x_k-x_{k-1}` and :math:`\Delta g=g_k-g_{k-1}`. A Barzilai-Borwein (BB) iteration is :math:`x_{k+1}=x_k-\alpha _kg_k` where the step size :math:`\alpha _k` is either
+
+    - :math:`\alpha _k^{LONG}=\frac{\Delta x\cdot\Delta x}{\Delta x\cdot\Delta g}`, or
+
+    - :math:`\alpha _k^{SHORT}=\frac{\Delta x \cdot\Delta g}{\Delta g \cdot\Delta g}`.
+    
+    Parameters
+    ----------
+    initial: float,
+        The step-size for the first iteration
+    mode: One of 'long', 'short' or 'alternate', default is 'short'. 
+        This calculates the step-size based on the LONG, SHORT or alternating between the two, starting with short. 
+    stabilisation_param: float or None, default is None 
```

Comment:
Unclear how behaviour is different from None and np.inf?  

I'm also unsure how it is decieded whether a positive constant is used or the minimum of three iterations, is that something the user specifies or the algorithm determines?

I think it might be more user-friendly to have stability switched on by default, ensuring convergence, then if the user wants, they can switch it off for potential further speed-up potentially sacrificing convergence but actively doing so at their own risk.  Is there a sensible default value that could be specified?

Reply:
Thanks Jakob, by default this is value is calculated automatically from the first three iterations. Otherwise, the user can pass a chosen constant or pass `np.inf` which is effectively turns the stabilisation off

Reply:
Thanks for the clarification.  Hmm, I think it is not intuitive that stabilization=None applies a (default) stabilization - None would seem to suggest the no stabilization is used, ie it is switched off.  I wonder how this could be made clearer. How about, in addition to float values, also allowing the string values 'default' and 'off', while not allowing None to be passed.